lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -483,11 +491,22 @@ private synchronized ErrorInfo waitForChange() {
          * ConnectionStateListener interface
          */
         @Override
-        public void onConnectionStateChanged(ConnectionStateChange state) {
-            synchronized(this) {
-                change = state;
-                notify();
+        public synchronized void onConnectionStateChanged(ConnectionStateChange state) {
+            change = state;
+            notify();
+        }
+
+        /**
+         * Remove this ConnectionWaiter as a connection listener.
+         */
+        private void close() {
+            // This method is explicitly not synchronized. There may be a case for this in the
+            // future, however its addition is designed to be lightweight with minimal impact.
+            if (closed) {
+                return;
             }
+            closed = true;
```

Comment:
I feel `closed=true` can be set after `connection.off(this);`, that way it's logically closed after removing it from connection

Reply:
Thanks for the suggestion, but I still prefer it this way around.